### ISSUE TITLE
Missing v-site exceptions with plugins

### ISSUE THREAD

**Issue by jthorton** [openforcefield/openff-interchange#919]
**Description**
I am trying to run some pure water simulations using the DEXP force field which keep crashing and this is likely due to missing v-site exceptions. In the example I expect 12 exceptions for a system of two water molecules each with a single v-site, corresponding to 3 normal exceptions between the H-O*2 and H-H and 3 extra exceptions between the site and H *2 and O. Then doubled for both sites. However it seems like the exceptions for the site in the first water are missing. I can confirm that this does not happen without plugins as `tip4p_fb.offxml` works as expected. 

**Reproduction**

```python
import openmm
from openff.toolkit import Molecule, ForceField, Topology
from openff.interchange import Interchange

water = Molecule.from_smiles('O')
water.generate_conformers()

ff = ForceField('de-force-1.0.2.offxml', load_plugins=True)
top = Topology.from_molecules([water, water])
interchange = Interchange.from_smirnoff(force_field=ff, topology=top)

system = interchange.to_openmm_system(False)
with open('water_system.xml', 'w') as out:
    out.write(openmm.XmlSerializer.serialize(system))
```

**Output**
expected output
```
<Exceptions>
        <Exception eps="0" p1="0" p2="1" q="0" sig="1"/>
        <Exception eps="0" p1="0" p2="2" q="0" sig="1"/>
        <Exception eps="0" p1="1" p2="2" q="0" sig="1"/>
        <Exception eps="0" p1="3" p2="4" q="0" sig="1"/>
        <Exception eps="0" p1="3" p2="5" q="0" sig="1"/>
        <Exception eps="0" p1="4" p2="5" q="0" sig="1"/>
        <Exception eps="0" p1="0" p2="6" q="0" sig="1"/>
        <Exception eps="0" p1="3" p2="7" q="0" sig="1"/>
        <Exception eps="0" p1="6" p2="1" q="0" sig="1"/>
        <Exception eps="0" p1="6" p2="2" q="0" sig="1"/>
        <Exception eps="0" p1="7" p2="4" q="0" sig="1"/>
        <Exception eps="0" p1="7" p2="5" q="0" sig="1"/>
</Exceptions>
```

dexp output
```
<Exceptions>
        <Exception eps="0" p1="0" p2="1" q="0" sig="0"/>
        <Exception eps="0" p1="0" p2="2" q="0" sig="0"/>
        <Exception eps="0" p1="1" p2="2" q="0" sig="0"/>
        <Exception eps="0" p1="3" p2="4" q="0" sig="0"/>
        <Exception eps="0" p1="3" p2="5" q="0" sig="0"/>
        <Exception eps="0" p1="4" p2="5" q="0" sig="0"/>
        <Exception eps="0" p1="3" p2="1" q="0" sig="0"/>
        <Exception eps="0" p1="3" p2="2" q="0" sig="0"/>
        <Exception eps="0" p1="7" p2="4" q="0" sig="0"/>
        <Exception eps="0" p1="7" p2="5" q="0" sig="0"/>
        <Exception eps="0" p1="0" p2="3" q="0" sig="0"/>
        <Exception eps="0" p1="3" p2="7" q="0" sig="0"/>
</Exceptions>
```

**Software versions**

- Which operating system and version are you using? mac
- How did you install Interchange? mamba install openff-interchange
- What is the output of running `conda list`?
```
openff-amber-ff-ports     0.0.4              pyhca7485f_0    conda-forge
openff-forcefields        2023.11.0          pyhca7485f_0    conda-forge
openff-interchange-base   0.3.22             pyhd8ed1ab_0    conda-forge
openff-models             0.1.2              pyhca7485f_0    conda-forge
openff-toolkit-base       0.15.2             pyhd8ed1ab_0    conda-forge
openff-units              0.2.1              pyh1a96a4e_0    conda-forge
openff-utilities          0.1.11             pyhd8ed1ab_0    conda-forge
smirnoff-plugins          2024.01.0          pyhd8ed1ab_0    conda-forge
```

**Comment by mattwthompson** [comment]
Your two exception blocks each have 12 exceptions - is it that the number of exceptions are wrong or that the wrong ones are listed? I think the latter, based on what I found below. I need to draw out these exceptions again, I think.

Looks like all the work that went into #906 only hit the `combine_nonbonded_forces=True` code path; if I try to isolate it to that variable I get what I think is the problematic behavior:

```python
In [1]: import openmm
   ...: from openff.toolkit import Molecule, ForceField, Topology
   ...: from openff.interchange import Interchange
   ...:
   ...: water = Molecule.from_smiles("O")
   ...: water.generate_conformers()
   ...:
   ...: ff = ForceField("tip4p_fb.offxml", load_plugins=True)
   ...: top = Topology.from_molecules([water, water])
   ...: interchange = Interchange.from_smirnoff(force_field=ff, topology=top)
   ...:
   ...: # system = interchange.to_openmm_system(False)
   ...: for combine in [True, False]:
   ...:     system = interchange.to_openmm_system(combine)
   ...:     with open(f"{combine}_water_system.xml", "w") as out:
   ...:         out.write(openmm.XmlSerializer.serialize(system))
```
In `False_water_system.xml`, the exceptions are self-consistent but wrong:
```

	<Forces>
		<Force cutoff="1" energy="4*epsilon*((sigma/r)^12-(sigma/r)^6); sigma=(sigma1+sigma2)/2; epsilon=sqrt(epsilon1*epsilon2); " forceGroup="0" method="0" name="vdW force" switchingDistance="-1" type="CustomNonbondedForce" useLongRangeCorrection="0" useSwitchingFunction="0" version="3">
			...
			<Exclusions>
				<Exclusion p1="0" p2="1"/>
				<Exclusion p1="0" p2="2"/>
				<Exclusion p1="1" p2="2"/>
				<Exclusion p1="3" p2="4"/>
				<Exclusion p1="3" p2="5"/>
				<Exclusion p1="4" p2="5"/>
				<Exclusion p1="3" p2="1"/>
				<Exclusion p1="3" p2="2"/>
				<Exclusion p1="7" p2="4"/>
				<Exclusion p1="7" p2="5"/>
				<Exclusion p1="0" p2="3"/>
				<Exclusion p1="3" p2="7"/>
			</Exclusions>
		</Force>
		<Force alpha="0" cutoff="1" dispersionCorrection="1" ewaldTolerance=".0005" exceptionsUsePeriodic="0" forceGroup="0" includeDirectSpace="1" ljAlpha="0" ljnx="0" ljny="0" ljnz="0" method="0" name="Electrostatics force" nx="0" ny="0" nz="0" recipForceGroup="-1" rfDielectric="78.3" switchingDistance="-1" type="NonbondedForce" useSwitchingFunction="0" version="4">
			<Exceptions>
				<Exception eps="0" p1="0" p2="1" q="0" sig="0"/>
				<Exception eps="0" p1="0" p2="2" q="0" sig="0"/>
				<Exception eps="0" p1="1" p2="2" q="0" sig="0"/>
				<Exception eps="0" p1="3" p2="4" q="0" sig="0"/>
				<Exception eps="0" p1="3" p2="5" q="0" sig="0"/>
				<Exception eps="0" p1="4" p2="5" q="0" sig="0"/>
				<Exception eps="0" p1="3" p2="1" q="0" sig="0"/>
				<Exception eps="0" p1="3" p2="2" q="0" sig="0"/>
				<Exception eps="0" p1="7" p2="4" q="0" sig="0"/>
				<Exception eps="0" p1="7" p2="5" q="0" sig="0"/>
				<Exception eps="0" p1="0" p2="3" q="0" sig="0"/>
				<Exception eps="0" p1="3" p2="7" q="0" sig="0"/>
			</Exceptions>
		</Force>
```
The second molecule's oxygen seems to be the center of these errors; the 

The contents of `True_water_system.xml` seem okay:
```
				<Exception eps="0" p1="0" p2="1" q="0" sig="1"/>
				<Exception eps="0" p1="0" p2="2" q="0" sig="1"/>
				<Exception eps="0" p1="1" p2="2" q="0" sig="1"/>
				<Exception eps="0" p1="3" p2="4" q="0" sig="1"/>
				<Exception eps="0" p1="3" p2="5" q="0" sig="1"/>
				<Exception eps="0" p1="4" p2="5" q="0" sig="1"/>
				<Exception eps="0" p1="6" p2="1" q="0" sig="1"/>
				<Exception eps="0" p1="6" p2="2" q="0" sig="1"/>
				<Exception eps="0" p1="7" p2="4" q="0" sig="1"/>
				<Exception eps="0" p1="7" p2="5" q="0" sig="1"/>
				<Exception eps="0" p1="0" p2="6" q="0" sig="1"/>
				<Exception eps="0" p1="3" p2="7" q="0" sig="1"/>
```
(In both cases the atom ordering is O, H, H, O, H, H, V, V)

I bet the problem is that atom indices aren't wired through `openff_openmm_particle_map`, since it's only the _second_molecule that sees this issue.

I think I need to add tests that at least
* Check that exceptions with virtual sites and without plugins are functionally the same whether or not the non-bonded force is split out
* Do the same with a plugin that modifies the vdW force (in principle this shouldn't change how exceptions are written).
* Do the same with a non-water molecule?

In general,
* It's really important to include more and more systems with multiple molecules with virtual sites in the topology; issues like these don't crop up at all in single-molecule topologies whether or not virtual sites are present
* The logic in creating OpenMM forces (just the non-bonded interactions) is becoming more and more unwieldy over time (up to [1100 lines](https://github.com/openforcefield/openff-interchange/blob/v0.3.22/openff/interchange/interop/openmm/_nonbonded.py)); I don't think we can afford to slim down any functionality, so it just needs more and more rigorous testing. (Bugs like these, even with the egg on my face, are integral in moving int his direction.)

**Comment by jthorton** [comment]
Ah sorry, to be clear it's that the wrong 12 exceptions are listed. I agree with the testing I was trying to get [these](https://github.com/jthorton/de-forcefields/pull/4) energy tests online and also hit this issue, maybe adding something similar where you calculate the energies and forces of a system with stored charges and positions might be useful for this case?

**Comment by mattwthompson** [comment]
Drawing it out helped - here's two waters, indexed how Interchange organizes particles with OpenMM, and with the exclusions listed out:

![image](https://github.com/openforcefield/openff-interchange/assets/7935382/72359038-2c65-4a1b-a446-d259444059cf)

This helps make it more clear which exceptions are garbage (3-0, 3-1, 3-2) and by extension which ones are missing. I think there's already a similar test with five-site waters with this sort of analysis, so it might be possible to just run that through with `combine_nonbonded_forces=False`.

**Comment by mattwthompson** [comment]
Okay, I found it. The update that ensured all virtual sites were added after all atoms did not cover the split force code path at all. Such a glaring omission can only sneak through via a lack of tests, so that'll be most of the work here.

I even found a simpler way, if accidental, to trigger this behavior. Since the indices are garbled and OpenMM doesn't (by default) add an exception between a pair of atoms where one is already present, just looking at the number of exceptions with multiple waters demonstrates this error. TIP5P, which happened to be the first test I looked at, should produce 10 intramolecular exceptions per water molecule, but a system with 5 waters has slightly fewer than 50 exceptions.
